monitor/src/main.rs

```diff
@@ -0,0 +1,289 @@
+use anyhow::bail;
+use chrono::{offset::Utc, DateTime, TimeZone};
+use contracts::GPv2Settlement;
+use futures::{stream, Stream, StreamExt};
+use shared::{transport::LoggingTransport, Web3};
+use std::time::Duration;
+use structopt::StructOpt;
+use web3::{
+    api::{EthFilter, Namespace},
+    types::{Block, BlockId, BlockNumber, Transaction, TransactionReceipt, H160},
+};
+
+const MIN_TRANSACTIONS: usize = 1;
+const MINUTES_BETWEEN_ALERTS: i64 = 10;
+const BLOCK_POLL_DURATION: Duration = Duration::from_secs(20);
+const NODE_RESPITE: Duration = Duration::from_secs(5);
+
+fn parse_unit(text: &str) -> anyhow::Result<f64> {
+    let num: f64 = text.parse()?;
+    if !(0.0..=1.0).contains(&num) {
+        bail!("not a number between zero and one");
+    }
+    Ok(num)
+}
+
+#[derive(Debug, StructOpt)]
+struct Arguments {
+    #[structopt(flatten)]
+    shared: shared::arguments::Arguments,
+
+    /// Which proportion of the previous transactions must fail before
+    /// triggering an alert in a specified time window.
+    /// Example: if the value is 0.1, then an alert is sent if more than 10% of
+    /// the previous transactions fail.
+    #[structopt(
+        long,
+        env = "REVERTED_TRANSACTIONS_ALERT_THRESHOLD",
+        default_value = "0.3",
+        parse(try_from_str = parse_unit),
+    )]
+    reverted_transaction_alert_threshold: f64,
+
+    /// How much back in the past to check for failed transactions in seconds.
+    #[structopt(
+        long,
+        env = "REVERTED_TRANSACTIONS_TIME_WINDOW",
+        default_value = "600",
+        parse(try_from_str = shared::arguments::duration_from_seconds),
+    )]
+    reverted_transaction_time_window: Duration,
+}
+
+async fn extract_transactions_to(
+    block: Block<Transaction>,
+    target: H160,
+    web3: &Web3,
+) -> impl Stream<Item = Option<TransactionReceipt>> + '_ {
+    let transactions = block
+        .transactions
+        .into_iter()
+        .filter(move |tx| tx.to == Some(target));
+    stream::unfold(transactions, move |mut txs| {
+        let eth = web3.eth();
+        async move {
+            if let Some(tx) = txs.next() {
+                let receipt = eth
+                    .transaction_receipt(tx.hash)
+                    .await
+                    .unwrap_or_else(|err| {
+                        tracing::warn!("error connecting to node: {:?}", err);
+                        None
+                    });
+                Some((receipt, txs))
+            } else {
+                None
+            }
+        }
+    })
+}
+
+// Note: this function does not manage uncle blocks. It could be that more than
+// one block is returned with the same number. Initialization may be subject to
+// race conditions where a few blocks might be skipped or duplicated.
+async fn get_blocks_from(
+    start: DateTime<Utc>,
+    web3: &Web3,
+) -> impl Stream<Item = anyhow::Result<Block<Transaction>>> + '_ {
```

Comment:
Haven't looked into the body in detail but maybe https://docs.rs/futures/0.3.15/futures/stream/trait.TryStreamExt.html trait functions could be used